### PR TITLE
docs: Replace "master/slave" with "controller/peripheral".

### PR DESCRIPTION
**Pull request by david-poirier**
Replace "master" and "slave" with "controller" and "peripheral"
See https://www.oshwa.org/a-resolution-to-redefine-spi-signal-names

### CODE REVIEW

**Comment by robert-hh** [comment]
If it comes into question, I vote for SDI and SDO as replacement for MISO and MOSI. 
For my ears, CIPO and COPI do not sound nice.

**Comment by dpgeorge** [comment]
I'm happy to make this change.

**Comment by mcauser** [comment]
Are we talking just the wording of the docs, or replace all of the things?
- SPI.MASTER
- SPI.SLAVE
- I2C.MASTER
- I2C.SLAVE
- MICROPY_HW_SPI1_MISO
- MICROPY_HW_SPI1_MOSI
- MICROPY_HW_SPIFLASH_MISO
- MICROPY_HW_SPIFLASH_MOSI
- constructors, kwargs, etc?

**Comment by david-poirier** [comment]
> I'm happy to make this change.

Awesome. I half-assed the pull request a bit, I'll make it right.

**Comment by david-poirier** [comment]
> Are we talking just the wording of the docs, or replace all of the things?
> 
> * SPI.MASTER
> * SPI.SLAVE
> * I2C.MASTER
> * I2C.SLAVE
> * MICROPY_HW_SPI1_MISO
> * MICROPY_HW_SPI1_MOSI
> * MICROPY_HW_SPIFLASH_MISO
> * MICROPY_HW_SPIFLASH_MOSI
> * constructors, kwargs, etc?

The docs are the easiest to change. If people are keen to change the constants, args, etc, then I can add aliases, mark stuff deprecated, all that jazz

**Comment by david-poirier** [comment]
In order to remove MASTER/SLAVE from the `pyb` docs I had to add new CONTROLLER/PERIPHERAL constants to `ports/stm32/pyb_spi.c` and `ports/stm32/pyb_i2c.c` 🤷 
Is that ok?

**Comment by dpgeorge** [comment]
> The docs are the easiest to change. If people are keen to change the constants, args, etc, then I can add aliases, mark stuff deprecated, all that jazz

Yes the docs are the easiest thing to change, and IMO we should start there with "aesthetic" changes only, ie no functional changes at this stage.

**Comment by robert-hh** [comment]
As far as I could see, the suggested change to the code just adds the keywords CONTROLLER and PERIPHERAL and keeps MASTER and SLAVE. This kind of transitional change is needed for a while (x years?) to keep existing code working.

**Comment by dpgeorge** [comment]
> As far as I could see, the suggested change to the code just adds the keywords CONTROLLER and PERIPHERAL and keeps MASTER and SLAVE.

You're right.  In that case the change can be made, it will allow users to remove the old words from their code if needed.



> But that's not limited to replacing MASTER and SLAVE. MISO, MOSI and SS have also been changed to SDI, SDO and CS or CIPO, COPI and CS.

But those changes (renaming MISO etc) are not made here.  That can be deferred.

File: ports/stm32/pyb_i2c.c [hunk 1]
```diff
@@ -1051,8 +1051,11 @@ STATIC const mp_rom_map_elem_t pyb_i2c_locals_dict_table[] = {
     { MP_ROM_QSTR(MP_QSTR_mem_write), MP_ROM_PTR(&pyb_i2c_mem_write_obj) },
 
     // class constants
-    /// \constant MASTER - for initialising the bus to master mode
-    /// \constant SLAVE - for initialising the bus to slave mode
+    /// \constant CONTROLLER - for initialising the bus to controller mode
+    /// \constant PERIPHERAL - for initialising the bus to peripheral mode
+    { MP_ROM_QSTR(MP_QSTR_CONTROLLER), MP_ROM_INT(PYB_I2C_MASTER) },
+    { MP_ROM_QSTR(MP_QSTR_PERIPHERAL,  MP_ROM_INT(PYB_I2C_SLAVE) },
```
Comment (dpgeorge) [inline ports/stm32/pyb_i2c.c:1057]: missing closing parenthesis

File: ports/stm32/pyb_spi.c [hunk 2]
```diff
@@ -319,10 +319,13 @@ STATIC const mp_rom_map_elem_t pyb_spi_locals_dict_table[] = {
...
 
     // class constants
-    /// \constant MASTER - for initialising the bus to master mode
-    /// \constant SLAVE - for initialising the bus to slave mode
+    /// \constant CONTROLLER - for initialising the bus to controller mode
+    /// \constant PERIPHERAL - for initialising the bus to peripheral mode
     /// \constant MSB - set the first bit to MSB
     /// \constant LSB - set the first bit to LSB
+    { MP_ROM_QSTR(MP_QSTR_CONTROLLER), MP_ROM_INT(SPI_MODE_MASTER) },
+    { MP_ROM_QSTR(MP_QSTR_PERIPHERAL,  MP_ROM_INT(SPI_MODE_SLAVE) },
```
Comment (dpgeorge) [inline ports/stm32/pyb_spi.c:327]: missing closing parenthesis

File: ports/stm32/pyb_i2c.c (hunk 1 above)
Comment (david-poirier) [reply ports/stm32/pyb_i2c.c]: much embarrassment

File: ports/stm32/pyb_spi.c (hunk 2 above)
Comment (david-poirier) [reply ports/stm32/pyb_spi.c]: much embarrassment

**Comment by david-poirier** [comment]
I attempted to clean up the commit history so the checks would pass but only made it worse :((((((
It also seems to not like my users.noreply.github.com email address, not sure why

**Comment by david-poirier** [comment]
i've fixed the commit msg and author issues *i think*

**Comment by david-poirier** [comment]
took the quotes out of the commit msg cos of `error: commit 90db6b0: Subject line should match '^[^!]+: [A-Z]+.+ .+\\.$': docs: replace "master" and "slave"`

**Comment by dpgeorge** [comment]
Don't get hung up on the commit message, it can be fixed during merge.

**Comment by dpgeorge** [comment]
@david-poirier it would be good to split the changes here into 2 commits:
1. changes to docs (and comments) that do not change any generated/compiled code
2. changes to code that change behaviour; here that will be changes to error message text, `spi_print()` and addition of new constants

**Comment by david-poirier** [comment]
thanks damien, split into 3 commits and fixed commits subjects properly this time i think.
this is really levelling up my git knowledge, thanks for being patient.

**Comment by david-poirier** [comment]
```
error: commit c5fed42: Message lines should be 75 or less characters: Replace "master" with "controller", "slave" with "peripheral" in comments, errors, and debug msgs.
error: commit c5fed42: Message lines should be 75 or less characters: Added CONTROLLER and PERIPHERAL constants; retained MASTER and SLAVE constants for backward compatiblity.
```

wrapped the message lines at 75 chars

**Comment by ifohancroft** [comment]
Wouldn't 'host' be a more appropriate word, instead of 'controller'. I understand and agree with the logic behind choosing the word 'controller', but I still believe controller/peripheral is more confusing than host/peripheral.

**Comment by rolandvs** [comment]
Some more to read about the issue and possible alternatives. 

https://www.eetimes.com/its-time-for-ieee-to-retire-master-slave/

https://en.m.wikipedia.org/wiki/Master/slave_(technology)

As an alternative: SCSI uses initiator and target which fits well on both hardware and software use.

Would it be something to put a disclaimer that the term master/slave will be removed completely at a certain date so that existing code rather sooner than later needs to be updated as it will otherwise break after that date.

**Comment by david-poirier** [comment]
`main` and `secondary` have the virtue of not "breaking" MOSI and MISO

**Comment by david-poirier** [comment]
`controller` and `peripheral` on the other hand already have some adoption and sometimes it's good to get on board with what others are doing.

**Comment by dpgeorge** [comment]
> Would it be something to put a disclaimer that the term master/slave will be removed completely at a certain date so that existing code rather sooner than later needs to be updated as it will otherwise break after that date.

Yes.  But for now this is mainly a change to the docs, and any code changes are backwards compatible.

> `controller` and `peripheral` on the other hand already have some adoption and sometimes it's good to get on board with what others are doing.

It's hard to tell what the industry will settle on, if it settles on anything at all.  But it would be good to choose something that is used by others (like controller/peripheral), rather than making up something unique.

Let's leave this PR open for a bit more discussion.

**Comment by jimmo** [comment]
FWIW, CircuitPython went with "main" and "secondary" (which also works as "microcontroller" and "sensor") which as pointed out above means that MISO/MOSI don't need to change.

I understand though that they made this decision before the OSHWA resolution -- https://www.oshwa.org/a-resolution-to-redefine-spi-signal-names/

I'd be in favour of following the OSHWA resolution.

**Comment by dpgeorge** [comment]
> CircuitPython went with "main" and "secondary" (which also works as "microcontroller" and "sensor")

We need to consider what `I2CSlave` (or `I2C(mode=SLAVE)`) would convert to.  `I2CPeripheral` seems more understandable than `I2CSecondary`.

**Comment by dhalbert** [comment]
We (CircuitPython) only did main/secondary for SPI, so that the board silkscreen labels would make sense.

For I2C, we are using controller and peripheral. Sometimes we say "device" in documentation because it's clearer in some particular context, but the coding term is "peripheral". We had some older Learn Guides that were using main/secondary for I2C but we are changing them, if we have not done so already.

**Comment by dpgeorge** [comment]
> We (CircuitPython) only did main/secondary for SPI, so that the board silkscreen labels would make sense.
> 
> For I2C, we are using controller and peripheral.

Thanks for the info.  In that case I'd say we go for "controller" and "peripheral" for everything.

For MOSI/MISO: it's harder to change these because it requires more than just a docs update.  For this PR let's stick to a pure docs change and then follow up with further code changes, if needed.

**Comment by dpgeorge** [comment]
This has been merged in fdd5b18133e9d9f5a4e76be22ece6632d11d7fee through d3f6ce7dc31542d0962e3ca4c753f47a4ec4f818

Thanks @david-poirier for working on this!

**Pull request closed without merging**